### PR TITLE
feat(tabs, segmented-control) DLT-3292 DLT-3293 DLT-3294 traveling indicator

### PR DESCRIPTION
**Pull request by francisrupert**
seinfeld-kramer

## :hammer_and_wrench: Type Of Change

- [x] Feature

## :book: Jira Ticket

- [DLT-3292](https://dialpad.atlassian.net/browse/DLT-3292) — Tabs: traveling indicator
- [DLT-3293](https://dialpad.atlassian.net/browse/DLT-3293) — Segmented Control: traveling indicator
- [DLT-3294](https://dialpad.atlassian.net/browse/DLT-3294) — Shared indicator animation composable
- [DLT-3291](https://dialpad.atlassian.net/browse/DLT-3291) — Add `--ttf-ease` token

## :book: Description

Adds a "traveling indicator" animation to DtTabGroup and DtSegmentedControl. When clicking a new tab or segmented control item, the selection indicator (underline, border, or background) visually slides from the old selection to the new one instead of instantly switching.

**What it does:**
- FLIP technique + Web Animations API (`Element.animate` on pseudo-elements) — no markup changes
- DtTabGroup: all four variants (default underline, outlined border, muted+outlined, muted active background)
- DtSegmentedControl: active background variant
- Shared `useIndicatorAnimation` composable + `indicatorAnimation.js` utility
- `showIndicatorTransition` prop on both components (default `true`) for consumer opt-out
- Duration/easing driven by CSS custom properties (`--tab-indicator-duration`, `--segmented-indicator-duration`, etc.)
- Respects `prefers-reduced-motion`
- Keyboard auto-activation (`activation-mode="auto"`) snaps instantly — no animation

**What it does not do:**
- No template/markup changes to existing components
- No changes to selection logic or event contracts
- No new dependencies

**Other changes in this PR:**
- New `--ttf-ease` token (`cubic-bezier(0.25, 0.1, 0.25, 1)`) + `.d-ttf` utility class
- Fix: `.ttf-out-quint` → `.d-ttf-out-quint` (missing `d-` prefix)

## :mag: For Reviewers

Test on the deploy preview:

1. **[Tabs](https://dialtone.dialpad.com/deploy-previews/pr-1192/components/tabs.html)** — click between tabs in each variant demo (Default, Borderless, Outlined, Muted, Muted+Outlined). Indicator should slide smoothly.
2. **[Segmented Control](https://dialtone.dialpad.com/deploy-previews/pr-1192/components/segmented-control.html)** — click between items. Background should slide.
3. **Keyboard** — tab into a group, use arrow keys. In manual mode (default), arrows move focus only; Enter/Space selects with animation. In auto mode, arrows snap instantly (no animation).
4. **Rapid clicks** — click tabs fast. Previous animation should cancel cleanly, no stuck transparent states.
5. **[Stress tests](https://dialtone.dialpad.com/deploy-previews/pr-1192/scratch.html)** — scroll to "Traveling Indicator Stress Test" section for edge cases: all sizes, spread modes, vertical, RTL, disabled, `showIndicatorTransition=false`, extreme width variance, many tabs, multiple instances.

## :package: Cross-Package Impact

| Package | Changes |
|---|---|
| dialtone-css | Animation CSS vars, `::before`/`::after` overlay rules, new `--ttf-ease` token |
| dialtone-vue | `tab_group.vue`, `segmented_control.vue`, new composable + utility |
| dialtone-documentation | Transition utility docs updated, scratch.md stress tests |

## :pencil: Checklist

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description.
- [x] I have reviewed my changes.
- [x] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

For all Vue changes:

- [x] I have added / updated unit tests.
- [x] I have validated components with a screen reader.
- [x] I have validated components keyboard navigation.

For all CSS changes:

- [x] I have used design tokens whenever possible.
- [x] I have considered how this change will behave on different screen sizes.
- [x] I have visually validated my change in light and dark mode.
- [x] I have used gap or flexbox properties for layout instead of margin whenever possible.

## :camera: Screenshots / GIFs / Screencasts

https://github.com/user-attachments/assets/85812238-443b-4aeb-8d35-ffd21d4dc5fe

https://github.com/user-attachments/assets/4313700c-d167-4b84-bede-24dec9bc62e2

[DLT-3292]: https://dialpad.atlassian.net/browse/DLT-3292?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DLT-3293]: https://dialpad.atlassian.net/browse/DLT-3293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DLT-3294]: https://dialpad.atlassian.net/browse/DLT-3294?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DLT-3291]: https://dialpad.atlassian.net/browse/DLT-3291?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Implements traveling indicator animation for tabs and segmented controls, smoothly sliding the selection indicator between items using the FLIP technique and Web Animations API. Introduces useIndicatorAnimation composable, showIndicatorTransition prop (default true), new --ttf-ease token, and CSS-driven duration/easing. Respects prefers-reduced-motion.

### CODE REVIEW

**Review by francisrupert — COMMENTED** [review]
Addressed both CodeRabbit findings:

1. **`'quint'` → `'out-quint'` in transition.md** — Fixed. The table now correctly renders `.d-ttf-out-quint` and `var(--ttf-out-quint)`.

2. **`.ttf-out-quint` rename** — Acknowledged. No internal usages of the old unprefixed class. Will note in release notes.

Thanks CodeRabbit!

File: packages/dialtone-css/lib/build/less/utilities/effects.less [hunk 1]
```diff
@@ -189,9 +189,10 @@
 .d-td300 { transition-duration: var(--td300) !important; }
 
 //  --  TRANSITION TIMING
+.d-ttf { transition-timing-function: var(--ttf-ease) !important; }
 .d-ttf-in-out { transition-timing-function: var(--ttf-in-out) !important; }
 .d-ttf-out { transition-timing-function: var(--ttf-out) !important; }
-.ttf-out-quint { transition-timing-function: var(--ttf-out-quint) !important; }
+.d-ttf-out-quint { transition-timing-function: var(--ttf-out-quint) !important; }
```
[the remark this answers is not in the document]
Comment (francisrupert) [reply packages/dialtone-css/lib/build/less/utilities/effects.less:195]: https://github.com/dialpad/dialtone/pull/1192#pullrequestreview-4094455632

File: apps/dialtone-documentation/docs/utilities/effects/transition.md [hunk 2]
```diff
@@ -76,9 +77,9 @@ Use `d-tp-{n}` change an what items within an element are transitioned.
         <th scope="row" class="d-code--sm d-docsite-code">.d-{{ i }}{{ d }}</th>
         <td class="d-code--sm">transition-duration: var(--td{{ d }}) !important;</td>
       </tr>
-      <tr v-else-if="i === 'ttf'" v-for="t in ['in-out', 'out', 'quint']">
-        <th scope="row" class="d-code--sm d-docsite-code">.d-{{ i }}-{{ t }}</th>
-        <td class="d-code--sm">transition-timing-function: var(--ttf-{{ t }}) !important;</td>
+      <tr v-else-if="i === 'ttf'" v-for="t in ['', 'in-out', 'out', 'quint']">
+        <th scope="row" class="d-code--sm d-docsite-code">.d-{{ i }}{{ t ? `-${t}` : '' }}</th>
+        <td class="d-code--sm">transition-timing-function: var(--ttf-{{ t || 'ease' }}) !important;</td>
```
[the remark this answers is not in the document]
Comment (francisrupert) [reply apps/dialtone-documentation/docs/utilities/effects/transition.md]: https://github.com/dialpad/dialtone/pull/1192#pullrequestreview-4094455632

**Comment by francisrupert** [comment]
Addressing the CodeRabbit review summary:

**Inline comments (both resolved):**

1. ✅ `'quint'` → `'out-quint'` in `transition.md` — fixed in c82b2e4.
2. ✅ `.ttf-out-quint` → `.d-ttf-out-quint` rename — already in the diff. No internal usages of the old class found.

**Outside diff comment (unit tests for `transitionIndicator`):**

Not adding. The animation code gracefully degrades in jsdom (`Element.animate` doesn't exist), and all 65 tab + 30 segmented control tests pass. The animation is purely visual — it doesn't change observable state (selection, events, focus). Testing spy calls on `animate` would be testing implementation details, not behavior.

**Nitpick (WAAPI feature detection):**

Already handled. Both components guard with `typeof newContext?.animate !== 'function'` / `typeof newEl.animate !== 'function'` before calling the API. That IS the WAAPI feature detection. Every browser that supports `Element.animate()` also supports pseudo-element animation and `translate`/`scale` transforms. Additional capability sniffing is unnecessary.

**Comment by francisrupert** [comment]
**Re: Codex review — multi-row tab wrapping**

Good catch. Fixed — we now detect when old and new elements are on different rows (cross-axis delta > 1px) and skip the animation entirely (instant switch). We don't want a diagonal transition, and wrapped tabs are an extreme edge case that ideally wouldn't occur in practice.

File: packages/dialtone-vue/components/segmented_control/segmented_control.vue [hunk 3]
```diff
@@ -226,14 +241,46 @@ function setFocus (value) {
...
+    }
+  }
+
   emit('update:modelValue', value);
   emit('change', value);
+
+  if (!oldRect || !canAnimate) return;
+
+  nextTick(() => {
+    const newEl = containerEl.querySelector(`[${SEGMENTED_CONTROL_DATA_VALUE_ATTR}="${value}"]`);
```
[the remark this answers is not in the document]
Comment (francisrupert) [reply packages/dialtone-vue/components/segmented_control/segmented_control.vue]: Good catch. Added `CSS.escape(value)` to prevent `querySelector` from throwing on values with special characters.

File: packages/dialtone-vue/common/utils/indicatorAnimation.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  state.hideNative = newEl.animate(
+    [hideProps, hideProps], { ...opts, fill: 'forwards' },
+  );
+  state.indicator = newEl.animate(
+    [{ translate: from, scale, ...indicatorExtra },
+      { translate: '0 0', scale: '1 1', ...indicatorExtra }],
+    { ...opts, pseudoElement },
+  );
+  state.indicator.onfinish = () => { state.hideNative?.cancel(); };
+}
```
[the remark this answers is not in the document]
Comment (francisrupert) [reply packages/dialtone-vue/common/utils/indicatorAnimation.js]: Already addressed — see [this comment](https://github.com/dialpad/dialtone/pull/1192#issuecomment-4230649793). The animation is purely visual and gracefully degrades (jsdom has no `Element.animate`). All 95 existing behavioral tests pass. Adding spy-based tests for `animate` calls would test implementation details, not behavior.

File: packages/dialtone-vue/components/segmented_control/segmented_control.vue [hunk 5]
```diff
@@ -226,14 +241,47 @@ function setFocus (value) {
...
+      oldBg = getComputedStyle(oldEl).backgroundColor;
+    }
+  }
+
   emit('update:modelValue', value);
   emit('change', value);
+
+  if (!oldRect || !canAnimate) return;
+
+  nextTick(() => {
```
Comment (braddialpad) [inline packages/dialtone-vue/components/segmented_control/segmented_control.vue]: Can we use async await instead?

File: packages/dialtone-css/lib/build/less/utilities/effects.less (hunk 1 above)
Comment (braddialpad) [inline packages/dialtone-css/lib/build/less/utilities/effects.less:195]: breaking change, but probably okay. I don't see any usages of this utility class. Only direct use of `var(--ttf-out-quint)`

File: packages/dialtone-vue/components/tab/tab_group.vue [hunk 6]
```diff
@@ -381,14 +400,60 @@ export default {
...
+
+      const tabsEl = this.$refs.tabs;
+      const oldTab = tabsEl.querySelector('[aria-selected="true"]');
+      const oldRect = oldTab?.getBoundingClientRect();
+
+      let oldStyle = null;
+      if (oldTab && (this.outlined || this.kind === 'muted')) {
+        const cs = getComputedStyle(oldTab);
+        oldStyle = { borderColor: cs.borderColor, borderWidth: cs.borderWidth, backgroundColor: cs.backgroundColor };
+      }
```
Comment (braddialpad) [inline packages/dialtone-vue/components/tab/tab_group.vue]: Could probably isolate this logic in the composable and use it in both components. Something like..

```                                                                                                                                                                                                                                                                                      
function snapshot (selector) {                                                                                                                                                                                                                                                                                
  if (!canAnimate()) return null;
  const containerEl = /* resolve from containerRef */;                                                                                                                                                                                                                                                        
  const el = containerEl?.querySelector(selector);                                                                                                                                                                                                                                                            
  if (!el) return null;                                                                                                                                                                                                                                                                                       
  return {                                                                                                                                                                                                                                                                                                    
    rect: el.getBoundingClientRect(),
    style: getComputedStyle(el),
    el,                                                                                                                                                                                                                                                                                                       
  };
}
```

File: packages/dialtone-vue/components/segmented_control/segmented_control.vue [hunk 7]
```diff
@@ -226,14 +241,47 @@ function setFocus (value) {
...
+    const escapedValue = typeof CSS !== 'undefined' && CSS.escape ? CSS.escape(value) : value;
+    const newEl = containerEl.querySelector(`[${SEGMENTED_CONTROL_DATA_VALUE_ATTR}="${escapedValue}"]`);
+    if (!newEl || typeof newEl.animate !== 'function') return;
```
Comment (braddialpad) [inline packages/dialtone-vue/components/segmented_control/segmented_control.vue]: It would be nice to keep all "browser support detection" stuff in one place (such as `typeof newEl.animate !== 'function'`). We do some  browser checks already in `cacheIndicatorConfig`, perhaps there?

File: packages/dialtone-vue/components/segmented_control/segmented_control.vue [hunk 8]
```diff
@@ -226,14 +241,47 @@ function setFocus (value) {
...
+    }
+  }
+
   emit('update:modelValue', value);
   emit('change', value);
+
+  if (!oldRect || !canAnimate) return;
+
+  nextTick(() => {
+    const escapedValue = typeof CSS !== 'undefined' && CSS.escape ? CSS.escape(value) : value;
```
Comment (braddialpad) [inline packages/dialtone-vue/components/segmented_control/segmented_control.vue]: These conditionals appear to be a bit overkill. From claude:

> CSS.escape has been standard since 2016. The Web Animations API used throughout this feature requires Chrome 84+, Firefox 83+, Safari 14+ — all of which have had CSS.escape for years. The fallback to a bare value would also produce a broken selector for any value with special
>    characters, making it worse than just erroring. Just call CSS.escape(value) directly.

File: packages/dialtone-vue/common/utils/indicatorAnimation.js
Comment (braddialpad) [inline packages/dialtone-vue/common/utils/indicatorAnimation.js:1]: Question on the general concept of this file. Is there some reason we are separating and exporting these functions? Is it planned to use these functions in places outside of the composable in the future, right now the only place they are used is within the composable which implies they should just be part of the composable itself. Would be a bit nicer to manage the state all within the composable.

**Review by braddialpad — COMMENTED** [review]
Looks and works very good. Mostly nit/optional comments here

File: packages/dialtone-vue/common/utils/indicatorAnimation.js
Comment (francisrupert) [reply packages/dialtone-vue/common/utils/indicatorAnimation.js:1]: Yeah, should have been in the composable from the start anyway. At this time, I don't really anticipate consumers needing the raw functions. I merged the utility functions into the composable, and all state and logic now lives in the composable. Remains internal, no export.

File: packages/dialtone-vue/components/segmented_control/segmented_control.vue (hunk 8 above)
Comment (francisrupert) [reply packages/dialtone-vue/components/segmented_control/segmented_control.vue]: Updated. If WAAPI is supported, CSS.escape is guaranteed. Guard dropped

File: packages/dialtone-vue/components/segmented_control/segmented_control.vue (hunk 7 above)
Comment (francisrupert) [reply packages/dialtone-vue/components/segmented_control/segmented_control.vue]: cacheIndicatorConfig is gone — merged into the composable. WAAPI detection and prefers-reduced-motion now checked at mount, stored as a single supported flag. The typeof newEl.animate guards in both components removed.

File: packages/dialtone-vue/components/tab/tab_group.vue (hunk 6 above)
Comment (francisrupert) [reply packages/dialtone-vue/components/tab/tab_group.vue]: Added a `snapshot(selector)` method to the composable that returns `{ rect, style, el }`. Both components now use `indicator.snapshot()` insteadof DOM queries.

File: packages/dialtone-vue/common/composables/useIndicatorAnimation.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+      : Math.abs(oldRect.top - newRect.top);
+    if (crossAxisDelta > 1) return;
+
+    // Center-to-center delta so scale from center aligns both edges
+    const delta = isVertical
+      ? (oldRect.top + oldRect.height / 2) - (newRect.top + newRect.height / 2)
+      : (oldRect.left + oldRect.width / 2) - (newRect.left + newRect.width / 2);
+    if (delta === 0) return;
+
+    cancel();
```
[the remark this answers is not in the document]
Comment (francisrupert) [reply packages/dialtone-vue/common/composables/useIndicatorAnimation.js]: Moved cancel() above the skip checks so previous animations are always cleared, even on wrapped-row or zero-delta early returns.

**Comment by braddialpad** [comment]
hmm looks like something broke in the segmented control after these changes.



Tabs still look good.

**Comment by francisrupert** [comment]
@braddialpad,

> ...something broke in the segmented control after these changes...

It also broke for Tabs in `outlined` or `muted` variants. `getComputedStyle()` returns a live object. By the time the animation runs **after** `nextTick`, the old element has lost its active/outlined class and the styles have changed. `snapshot()` now copies the values into a plain object at capture time. 

The default `::after` of Tabs underline variant wasn't affected since it only uses oldRect, not style properties.

This was messed up when moving to just composable.

File: packages/dialtone-css/lib/build/less/components/tabs.less [hunk 10]
```diff
@@ -103,6 +107,7 @@
 
     &__item {
         text-align: start;
+        transition: none;
```
Comment (braddialpad) [inline packages/dialtone-css/lib/build/less/components/tabs.less:110]: This `transition: none` overrides the dt-button's built in transition, causing the mouseover style to "flicker" while the transition is happening. I'm thinking this probably isn't intended behaviour?

Image

**Comment by braddialpad** [comment]
> It also broke for Tabs in `outlined` or `muted` variants. `getComputedStyle()` returns a live object. By the time the animation runs **after** `nextTick`, the old element has lost its active/outlined class and the styles have changed. `snapshot()` now copies the values into a plain object at capture time.
> 
> The default `::after` of Tabs underline variant wasn't affected since it only uses oldRect, not style properties.
> 
> This was messed up when moving to just composable.

Looks good now, thanks

File: packages/dialtone-css/lib/build/less/components/tabs.less (hunk 10 above)
Comment (francisrupert) [reply packages/dialtone-css/lib/build/less/components/tabs.less:110]: Yeah, intentional. I had to do that to minimize an even more distracting "flicker" as it transitions from the `:active` background color to no background color after you click/select the item before the traveling item reaches the target. It's not super ideal. I'm punting on this for now, but I'm thinking of not having a background color on hover/active on a non-active item for Tabs and Segmented Control, and adjust the text color instead on hover/active. Will keep an eye on that and keep that in back pocket.

Comment (braddialpad) [reply packages/dialtone-css/lib/build/less/components/tabs.less:110]: alright np if intentional then.

**Review by braddialpad — APPROVED** [review]
